### PR TITLE
Configurable range for auto scale

### LINKED ISSUE #135: Minimum scale would be useful
Hi!

I've been a very happy user of way-displays for over a year already. It's the best way of configuring displays I've seen on any platform.

Today I found the first display that I kind of have trouble with: a 27" full HD one. Since the pixels are really big, the autoscale puts me at 0.875 making everything blurry. 

```
I [12:31:01] HDMI-A-1 Changing:
I [12:31:01]   from:
I [12:31:01]     (no mode)
I [12:31:01]     VRR:       off
I [12:31:01]      (disabled)
I [12:31:01]   to:
I [12:31:01]     mode:      1920x1080@60Hz (60,000mHz) (preferred)
I [12:31:01]      (enabled)
I [12:31:02] 
I [12:31:02] Changes successful
I [12:31:02] 
I [12:31:02] HDMI-A-1 Changing:
I [12:31:02]   from:
I [12:31:02]     scale:     1.000 (0.844)
I [12:31:02]     position:  0,0
I [12:31:02]     mode:      1920x1080@60Hz (60,000mHz) (preferred)
I [12:31:02]     VRR:       off
I [12:31:02]   to:
I [12:31:02]     VRR:       on
I [12:31:02] 
I [12:31:02] HDMI-A-1: Cannot enable VRR, display or compositor may not support it.
I [12:31:02] 
I [12:31:02] eDP-1 Changing:
I [12:31:02]   from:
I [12:31:02]     scale:     1.250 (1.682)
I [12:31:02]     position:  0,0
I [12:31:02]     mode:      1920x1200@60Hz (60,001mHz) (preferred)
I [12:31:02]     VRR:       on
I [12:31:02]   to:
I [12:31:02]     position:  329,1234
I [12:31:02] 
I [12:31:02] HDMI-A-1 Changing:
I [12:31:02]   from:
I [12:31:02]     scale:     1.000 (0.844)
I [12:31:02]     position:  0,0
I [12:31:02]     mode:      1920x1080@60Hz (60,000mHz) (preferred)
I [12:31:02]     VRR:       off
I [12:31:02]   to:
I [12:31:02]     scale:     0.875
I [12:31:02] 
I [12:31:02] Changes successful
```

Of course, a simple `way-displays --set SCALE HDMI-A-1 1.0` fixes the issue. But I don't foresee any situation where I'd like a scale less than 1.0, so I'd configure it if I could.

If you think this is a good idea I could implement it and make a PR.

### PR DESCRIPTION
**Pull request by dancek**
Scales below 1.0 are typically unuseful, so make 1.0 a minimum auto scale by default.

Add configuration options for minimum and maximum auto scale, in case someone actually wants scales under 1.0 or wishes to limit the maximum scale.

Fixes #135.

### CODE REVIEW

**Comment by dancek** [comment]
I decided to make it configurable while I'm at it, and also add an optional maximum. (If I had access to multiple slightly over 200dpi displays I might well want to limit the maximum to 2.0.)

There's no command-line interface for adjusting the range as that feels like a rare operation.

**Comment by alex-courtis** [comment]
> I decided to make it configurable while I'm at it, and also add an optional maximum. (If I had access to multiple slightly over 200dpi displays I might well want to limit the maximum to 2.0.)

Many thanks. We'll likely need max at some stage...

> There's no command-line interface for adjusting the range as that feels like a rare operation.

That's fine, it would indeed be rare.

File: src/head.c [hunk 1]
```diff
@@ -154,8 +154,22 @@ wl_fixed_t head_auto_scale(struct Head *head) {
 	// round the dpi to the nearest 12, so that we get a nice even wl_fixed_t
 	long dpi_quantized = (long)(dpi / 12 + 0.5) * 12;
 
+	// convert min and max to quantized dpi
+	long dpi_min = 12 * (long)(min * 8 + 0.9999f);
```
Comment (alex-courtis) [inline src/head.c:158]: Why the 0.9999f ? Is it to do with the double parameters? Could they be floats?

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Many thanks, love your work.

Testing:

cli / log
- [ ] shows in -g
- [x] shows in -y -g

min
- [ ] default - when min or max not present scale is always set to 1
- [x] above
- [x] below
- [x] bad
- [x] 0

max
- [ ] default - when min or max not present scale is always set to 1
- [x] above
- [x] below
- [x] 0

other
- [x] specific scale
- [x] scaling off
- [x] auto scaling off
- [x] change cfg

odd
- [ ] min > max - not quite predictable

File: doc/CONFIGURATION.md [hunk 2]
```diff
@@ -104,6 +104,22 @@ This may be disabled and scale 1 will be used, unless a `SCALE` has been specifi
...
+
+Set minimum value for auto scaling. The default is 1.0.
+
+```yaml
+AUTO_SCALE_MIN: 1.0
+```
+
+### AUTO_SCALE_MAX
+
+Set maximum value for auto scaling. Values below 1.0 are ignored.
```
Comment (alex-courtis) [inline doc/CONFIGURATION.md:117]: Can we just accept values below 1.0? There might be a use case...

**Comment by dancek** [comment]
Thanks for a really good review! I'll work on these when I have time (hopefully in a couple of days) and get back to you!

File: src/head.c (hunk 1 above)
Comment (dancek) [reply src/head.c:158]: This was a quick-n-dirty implementation of `ceil`, ie. the value needs to be rounded up. I'll write something smarter instead.

File: doc/CONFIGURATION.md (hunk 2 above)
Comment (dancek) [reply doc/CONFIGURATION.md]: Yeah, I changed that.

File: src/head.c (hunk 1 above)
Comment (dancek) [reply src/head.c]: Changed to ceil() from math.h.

**Comment by dancek** [comment]
> cli / log
> 
> * [ ]  shows in -g

Added.

> min
> 
> * [ ]  default - when min not present scale is always set to 1

Actually this was when max wasn't set, due to accidentally using the default min value as the default max value too. Fixed.

> odd
> 
> * [ ]  min > max - not quite predictable

Fixed by disabling it in this case.

---

Give it another look with these fixes. I'll be happy to fix any further issues.

**Review by alex-courtis — APPROVED** [review]
Fantastic, all working nicely:

cli / log
- [x] shows in -g
- [x] shows in -y -g

min
- [x] default - when min not present scale is always set to 1 - fixed
- [x] above
- [x] below
- [x] bad
- [x] 0

max
- [x] default
- [x] above
- [x] below
- [x] bad
- [x] 0
- [x] 0.1 - goes to calculated

min > max
- [x] uses min

other
- [x] specific scale
- [x] scaling off
- [x] auto scaling off
- [x] change cfg

**Comment by alex-courtis** [comment]
Many thanks for your contribution!

Can you wait for a release? I'd like to get #121 in as well...

**Comment by dancek** [comment]
I'm in no hurry, take your time!

**Pull request merged**